vulnerabilities/import_runner.py

```diff
@@ -109,169 +111,125 @@ def run(self, cutoff_date: datetime.datetime = None) -> None:
-                    **dataclasses.asdict(ref)))
+                if vuln_created or pkg_created:
+                    bulk_create_vuln_pkg_refs.add(pkg_vuln_ref)
+                    # insert all references of vuln + pkg
 
-    else:
-        # build indexes of existing references to either update or create new ones
-        # and avoid creating duplicates
+                else:
+                    # insert only if it is new vuln-pkg references
+                    existing_ref = get_vuln_pkg_refs(vuln, pkg)
+                    if not existing_ref:
+                        bulk_create_vuln_pkg_refs.add(pkg_vuln_ref)
 
-        existing_vuln_references = models.VulnerabilityReference.objects.filter(vulnerability=vuln)
-        refs_by_id_url = {
-            (ref.id, ref.url): ref for ref in existing_vuln_references if ref.id and ref.url}
-        refs_by_id = {ref.id: ref for ref in existing_vuln_references if ref.id}
-        refs_by_url = {ref.url: ref for ref in existing_vuln_references if ref.url}
+                    else:
+                        # This handles conflicts between existing datq and obtained data
+                        if existing_ref[0].is_vulnerable != pkg_vuln_ref.is_vulnerable:
+                            handle_conflicts([existing_ref[0], pkg_vuln_ref.to_model_object()])
+                            existing_ref.delete()
 
-        for adv_ref in advisory.vuln_references:
-            if adv_ref.reference_id and adv_ref.url:
-                existing_ref = refs_by_id_url.get((adv_ref.reference_id, adv_ref.url))
+    models.VulnerabilityReference.objects.bulk_create(
+        [i.to_model_object() for i in bulk_create_vuln_refs]
+    )
 
-            elif adv_ref.reference_id:
-                existing_ref = refs_by_id.get(adv_ref.reference_id)
+    # find_conflicting_relations handles in-memory conflicts
+    conflicts = find_conflicting_relations(bulk_create_vuln_pkg_refs)
 
-            elif adv_ref.url:
-                existing_ref = refs_by_url.get(adv_ref.url)
+    models.PackageRelatedVulnerability.objects.bulk_create(
+        [i.to_model_object() for i in bulk_create_vuln_pkg_refs if i not in conflicts]
+    )
 
-            if existing_ref:
-                # update ref as needed
-                if not adv_ref.reference_id or not adv_ref.url:
-                    continue
+    handle_conflicts([i.to_model_object() for i in conflicts])
 
-                if adv_ref.reference_id and not existing_ref.reference_id:
-                    existing_ref.reference_id = adv_ref.reference_id
 
-                if adv_ref.url and not existing_ref.url:
-                    existing_ref.url = adv_ref.url
+def find_conflicting_relations(relations):
 
-                existing_ref.save()
+    relation_tuples = [(rel.package, rel.vulnerability) for rel in relations]
+    relation_counter = Counter(relation_tuples).most_common()
+    conflicts = set()
+    for rel, count in relation_counter:
+        if count < 2:
```

Comment:
What is this `2`? The way I read this, you are trying to get a rel (which should should likely write as `for (vuln, pack), count` rather than `for rel, count`)... which as of the most common has only a single occurrence (meaning there could be more common ones we skip).... then you are create two objects `PackageRelatedVulnerabilityInserter` one is_vulnerable and one not is_vulnerable... Can you elaborate what you are trying to do there?

Reply:
@pombredanne  Sorry this is  cryptic and needs some doc strings.

The point of the function `find_conflicting_relations` is to detect entries like `(pkg=foo, vuln=bar, is_vulnerable=True)` and `(pkg=foo, vuln=bar, is_vulnerable=False)` from EXISTING TOGETHER in the `db insertion que`.

To do this : 

1. Take in  the `db insertion que` which is a SET of `PackageRelatedVulnerabilityInserter` this is basically `(pkg, vuln, is_vulnerable)`

2. Create a list of tuples from `PackageRelatedVulnerabilityInserter` by chopping off the `is_vulnerable` flag. These tuples are of format `(pkg, vuln)`

3. Now make a Counter object from this list of tuples.  Now here's the key idea , for a count of a particular `(pkg, vuln)` to be `>=2` it needs to have `>=2`  `PackageRelatedVulnerabilityInserter` objects to be instantiated from in the first place. 
 Since `db insertion que` is a SET, for this to be true it implies that those   `PackageRelatedVulnerabilityInserter` objects would be `(pkg, vuln, True)` and  `(pkg, vuln, False)`  which is conflicting data.

4. We then recreate this conflicting data and bubble it up.

Also note that :

`relation_counter = Counter(relation_tuples).most_common()`
which means while iterating we encounter most common elements first. This implies `count(nth item) >= count((n+1)th item) `. If count is less 2 than all the entries ahead don't have conflicting entries and we halt the search.

Reply:
Thanks! this make sense. I think there may be ways to improve this in the future, but that's a perfect explanation. Adding the comments in https://github.com/nexB/vulnerablecode/pull/255/commits/7f1f476cf003ae975134181e0bc117368c6f1e22 is A-OK.